File: integrations/aws/provider.py
from __future__ import annotations

from dataclasses import dataclass
from typing import Protocol


@dataclass(frozen=True)
class AwsIdentity:
    account_id: str
    account_name: str
    role_arn: str
    region: str


class AwsProvider(Protocol):
    async def identity(self) -> AwsIdentity:
        """Return the active AWS identity used by read-only tools."""


class MockAwsProvider:
    def __init__(self, region: str = "us-east-1") -> None:
        self._identity = AwsIdentity(
            account_id="000000000000",
            account_name="simulated-platform-prod",
            role_arn="arn:aws:iam::000000000000:role/mock-readonly-aiops-agent",
            region=region,
        )

    async def identity(self) -> AwsIdentity:
        return self._identity

    async def accounts(self) -> list[AwsIdentity]:
        return [
            self._identity,
            AwsIdentity(
                account_id="111111111111",
                account_name="simulated-shared-services",
                role_arn="arn:aws:iam::111111111111:role/mock-readonly-aiops-agent",
                region=self._identity.region,
            ),
            AwsIdentity(
                account_id="222222222222",
                account_name="simulated-platform-staging",
                role_arn="arn:aws:iam::222222222222:role/mock-readonly-aiops-agent",
                region="us-west-2",
            ),
        ]


class AssumeRoleAwsProvider:
    """Interface placeholder for future STS AssumeRole implementation.

    Production code should configure an external role ARN and create short-lived
    read-only sessions. No credentials are stored in this repository.
    """

    def __init__(self, role_arn: str, region: str) -> None:
        self.role_arn = role_arn
        self.region = region

    async def identity(self) -> AwsIdentity:
        raise NotImplementedError("Real AWS AssumeRole provider is intentionally not wired in MVP.")


def get_aws_provider() -> AwsProvider:
    return MockAwsProvider()

File: src/integrations/aws/provider.ts
export type AwsIdentity = {
  accountId: string;
  accountName: string;
  roleArn: string;
  region: string;
};

export interface AwsProvider {
  identity(): Promise<AwsIdentity>;
  accounts(): Promise<AwsIdentity[]>;
}

export class MockAwsProvider implements AwsProvider {
  private readonly primary: AwsIdentity;

  constructor(region = "us-east-1") {
    this.primary = {
      accountId: "000000000000",
      accountName: "simulated-platform-prod",
      roleArn: "arn:aws:iam::000000000000:role/mock-readonly-aiops-agent",
      region,
    };
  }

  async identity(): Promise<AwsIdentity> {
    return this.primary;
  }

  async accounts(): Promise<AwsIdentity[]> {
    return [
      this.primary,
      {
        accountId: "111111111111",
        accountName: "simulated-shared-services",
        roleArn: "arn:aws:iam::111111111111:role/mock-readonly-aiops-agent",
        region: this.primary.region,
      },
      {
        accountId: "222222222222",
        accountName: "simulated-platform-staging",
        roleArn: "arn:aws:iam::222222222222:role/mock-readonly-aiops-agent",
        region: "us-west-2",
      },
    ];
  }
}

export class AssumeRoleAwsProvider implements AwsProvider {
  constructor(
    private readonly roleArn: string,
    private readonly region: string,
  ) {}

  async identity(): Promise<AwsIdentity> {
    throw new Error(
      `Real AWS AssumeRole provider is intentionally not wired in MVP for ${this.roleArn} in ${this.region}.`,
    );
  }

  async accounts(): Promise<AwsIdentity[]> {
    return [await this.identity()];
  }
}

export function getAwsProvider(): AwsProvider {
  return new MockAwsProvider(process.env.AWS_REGION ?? "us-east-1");
}
